plugins/parquet/parquet.cpp

```diff
@@ -6,511 +6,326 @@
-  auto event_col = rb->GetColumnByName("event");
-  auto schema_metadata = rb->schema()->GetFieldByName("event")->metadata();
-  auto event_rb = arrow::RecordBatch::FromStructArray(event_col).ValueOrDie();
-  return event_rb->ReplaceSchemaMetadata(schema_metadata);
-}
-
-/// Create multiple table slices for a record batch, splitting at `max_slice_size`
-std::vector<table_slice>
-create_table_slices(const std::shared_ptr<arrow::RecordBatch>& rb,
-                    int64_t max_slice_size) {
-  auto final_rb = unwrap_record_batch(rb);
-  auto time_col = rb->GetColumnByName("import_time");
-  auto slices = std::vector<table_slice>{};
-  slices.reserve(rb->num_rows() / max_slice_size + 1);
-  auto schema = type::from_arrow(*final_rb->schema());
-  for (int64_t offset = 0; offset < rb->num_rows(); offset += max_slice_size) {
-    auto rb_sliced = final_rb->Slice(offset, max_slice_size);
-    auto& slice = slices.emplace_back(rb_sliced, schema);
-    slice.import_time(
-      derive_import_time(time_col->Slice(offset, max_slice_size)));
-    slice.offset(detail::narrow_cast<id>(offset));
+  auto input_file = as_arrow_file(std::move(parquet_chunk));
+  auto parquet_reader_properties
+    = ::parquet::ReaderProperties(arrow::default_memory_pool());
+  parquet_reader_properties.enable_buffered_stream();
+  std::unique_ptr<::parquet::arrow::FileReader> out_buffer;
+  auto arrow_reader_properties = ::parquet::ArrowReaderProperties();
+
+  // 2^16 is the largest batch size allowed in a table slice
+  arrow_reader_properties.set_batch_size(65536);
+  try {
+    auto input_buffer = ::parquet::ParquetFileReader::Open(
+      std::move(input_file), parquet_reader_properties);
+    ::arrow::Status arrow_file_reader_status
+      = ::parquet::arrow::FileReader::Make(arrow::default_memory_pool(),
+                                           std::move(input_buffer),
+                                           arrow_reader_properties,
+                                           &out_buffer);
+    if (!arrow_file_reader_status.ok()) {
+      diagnostic::error("{}", arrow_file_reader_status.ToString())
+        .emit(ctrl.diagnostics());
+      co_return;
+    }
+  } catch (const ::parquet::ParquetInvalidOrCorruptedFileException& err) {
+    diagnostic::error("invalid or corrupted parquet file: {}", err.what())
+      .emit(ctrl.diagnostics());
+    co_return;
+  }
+  std::shared_ptr<::arrow::RecordBatchReader> rb_reader;
+  auto record_batch_reader_status
+    = out_buffer->GetRecordBatchReader(&rb_reader);
+  if (!record_batch_reader_status.ok()) {
+    diagnostic::error("{}", record_batch_reader_status.ToString())
+      .note("failed create record batches from input data")
+      .emit(ctrl.diagnostics());
+    co_return;
+  }
+  for (arrow::Result<std::shared_ptr<arrow::RecordBatch>> maybe_batch :
+       *rb_reader) {
+    if (!maybe_batch.ok()) {
+      diagnostic::error("{}", maybe_batch.status().ToString())
```

Comment:
Needs to be ToStringWithoutContextLines here and above, whenever we're creating a diagnostic